manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java

```diff
@@ -68,20 +90,26 @@ public T handleWork(Work work) {
         if (PROVISIONING_STARTED.contains(managedResource.getStatus())) {
             try {
                 updated = createDependencies(work, managedResource);
-                complete = isProvisioningComplete(updated);
             } catch (Exception e) {
                 LOGGER.error(String.format("Failed to create dependencies for resource of type '%s' with id '%s'.", work.getType(), id), e);
                 // Something has gone wrong. We need to retry.
                 workManager.rescheduleAfterFailure(work);
+                recordError(work, e);
+                updated = load(id);
+            } finally {
+                complete = isProvisioningComplete(updated);
             }
         } else if (DEPROVISIONING_STARTED.contains(managedResource.getStatus())) {
             try {
                 updated = deleteDependencies(work, managedResource);
-                complete = isDeprovisioningComplete(updated);
             } catch (Exception e) {
                 LOGGER.info(String.format("Failed to delete dependencies for resource of type '%s' with id '%s'.", work.getType(), id), e);
                 // Something has gone wrong. We need to retry.
                 workManager.rescheduleAfterFailure(work);
+                recordError(work, e);
+                updated = load(id);
```

Comment:
@manstis I'm not sure the logic here feels right. If I read the code correctly, we could end up calling both `rescheduleAfterFailure` and `reschedule` if the `complete` boolean is set to `false` in the `finally` block.

I'm also not sure we need to record the intermediate failure status as it would only be visible on the resource in between retries. As a suggestion, I think it would be absolutely fine to record the final failure status_code e.g timed out or max retries exceeded.

Reply:
I agree. Good spot. Refactored.